core/src/impl/Kokkos_FunctorAnalysis.hpp

```diff
@@ -655,6 +655,11 @@ struct FunctorAnalysis {
                         detected_volatile_join_no_tag<F>::value)>>
       : public has_volatile_join_no_tag_function<F> {
     enum : bool { value = true };
+#ifndef KOKKOS_ENABLE_DEPRECATED_CODE_3
+    static_assert(Impl::dependent_false_v<F>,
```

Comment:
I would have gone for `std::is_void_v` but I don't necessarily have an issue with your helper variable template since it convey the intent.